src/main/java/io/openliberty/tools/intellij/lsp4jakarta/lsp4ij/codeAction/proposal/quickfix/RemoveParamAnnotationQuickFix.java

```diff
@@ -57,44 +65,66 @@ public List<? extends CodeAction> getCodeActions(JavaCodeActionContext context,
             }
             if (!annotationsToRemove.isEmpty()) {
                 // Create label
-                final StringBuilder sb = new StringBuilder();
-                // Java annotations in comma delimited list, assume that is ok.
-                sb.append("'@").append(getShortName(annotationsToRemove.get(0))).append("'");
-                for (int j = 1; j < annotationsToRemove.size(); ++j) {
-                    sb.append(", '@").append(getShortName(annotationsToRemove.get(j))).append("'");
-                }
-                String label = Messages.getMessage("RemoveTheModifierFromParameter", sb.toString(), parameter.getName().toString());
-                // Remove annotations
-                removeAnnotations(diagnostic, context.copy(), codeActions, i, label, annotationsToRemove);
+                String label = getLabel(parameter, annotationsToRemove);
+                Map<String, Object> extendedData = new HashMap<>();
+                extendedData.put(ANNOTATION_TO_REMOVE, annotationsToRemove);
+                extendedData.put(INDEX_KEY, i);
+                codeActions.add(JDTUtils.createCodeAction(context, diagnostic, label, getParticipantId(), extendedData));
             }
         }
         return codeActions;
     }
+    private String getLabel(PsiParameter parameter,List<String> annotationsToRemove) {
+        final StringBuilder sb = new StringBuilder();
+        // Java annotations in comma delimited list, assume that is ok.
+        sb.append("'@").append(getShortName(annotationsToRemove.get(0))).append("'");
+        for (int j = 1; j < annotationsToRemove.size(); ++j) {
+            sb.append(", '@").append(getShortName(annotationsToRemove.get(j))).append("'");
+        }
+        return Messages.getMessage("RemoveTheModifierFromParameter", sb.toString(), parameter.getName().toString());
+    }
 
-    protected void removeAnnotations(Diagnostic diagnostic, JavaCodeActionContext context,
-                                     List<CodeAction> codeActions, int parameterIndex,
-                                     String label, List<String> annotationsToRemove) {
+     @Override
+     public CodeAction resolveCodeAction(JavaCodeActionResolveContext context) {
+         final PsiElement node = context.getCoveredNode();
+         final CodeAction toResolve = context.getUnresolved();
+         final PsiClass parentType = getBinding(node);
+         final PsiMethod method = PsiTreeUtil.getParentOfType(node, PsiMethod.class);
+         CodeActionResolveData data = (CodeActionResolveData) toResolve.getData();
+         List<String> annotationsToRemove = new ArrayList<>();
+         int paramIndex = 0;
 
-        final PsiElement node = context.getCoveredNode();
-        final PsiClass parentType = getBinding(node);
-        final PsiMethod method = PsiTreeUtil.getParentOfType(node, PsiMethod.class);
+         if (data.getExtendedDataEntry(ANNOTATION_TO_REMOVE) instanceof List) {
```

Comment:
If this check fails then we have a serious error. Please add an else statement that logs a message like "The CodeActionResolveData was corrupted somewhere in the LSP layer. Unable to resolve code action." and then return null from resolveCodeAction(). I think null will be passed back to the language server and be ignored but need to test. 

This also applies to the INDEX_KEY case. Log the same message and return null. It may be possible to combine these two checks. 

Now you can remove the default values for annotationsToRemove and paramIndex because they will not be used.